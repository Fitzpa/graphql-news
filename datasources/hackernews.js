import { RESTDataSource } from 'apollo-datasource-rest';

export class HackerNewsAPI extends RESTDataSource {
  constructor() {
    super();
    this.baseURL = 'https://hacker-news.firebaseio.com/v0/';
  }

  articleReducer({ id, by, url, time, title } = {}) {
    return {
      id: `hn-${id}`,
      title,
      author: by,
      url,
      time,
      source: 'HackerNews',
    };
  }

  async getAllArticleIds() {
    const result = await this.get('topstories.json');
    return result;
  }

  async getArticle(articleId) {
    const result = await this.get(`item/${articleId}.json`);
    return this.articleReducer(result);
  }

  getArticlesByIds(articleIds) {
    return Promise.all(articleIds.map(articleId => this.getArticle(articleId)));
  }

  async getAllArticles() {
    let articleIds = await this.getAllArticleIds();
    articleIds = articleIds.slice(0, 100);
    return Promise.all(articleIds.map(articleId => this.getArticle(articleId)));
  }
}