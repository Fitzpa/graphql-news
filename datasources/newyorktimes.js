import { RESTDataSource } from 'apollo-datasource-rest';

export class NewYorkTimesAPI extends RESTDataSource {
  articleReducer({ id, byline, url, published_date, title } = {}) {
    return {
      id: `nyt-${id}`,
      title,
      author: byline,
      url,
      time: published_date,
      source: 'New York Times',
    };
  }

  async getAllArticles() {
    const result = await this.get(
      'https://api.nytimes.com/svc/mostpopular/v2/viewed/1.json?api-key=Jg1WQqSCu08PZKrQyqPGNN7F7FCYVnxp'
    );
    return result?.results?.map(article => this.articleReducer(article));
  }
}
