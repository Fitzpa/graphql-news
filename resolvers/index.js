export const resolvers = {
  Query: {
    allArticles: (_, __, { dataSources }) =>
      Promise.all(
        Object.keys(dataSources).map(source =>
          dataSources[source].getAllArticles()
        )
      ).then(result => result.reduce((acc, data) => acc.concat(data), [])),

    allArticlesBySource: (_, { source }, { dataSources }) =>
      dataSources[source].getAllArticles(),

    articleBySource: (_, { id, source }, { dataSources }) =>
      dataSources[source].getArticle(id, source),

    articlesBySource: (_, { ids, source }, { dataSources }) =>
      dataSources[source].getArticlesByIds(ids),
  },
};
