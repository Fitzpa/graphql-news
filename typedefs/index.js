import { gql } from 'apollo-server';

export const typeDefs = gql`
  type Query {
    allArticles: [Article!]!
    allArticlesBySource(source: String!): [Article!]!
    articleBySource(id: ID!, source: String!): Article
    articlesBySource(ids: [Int!]!, source: String!): [Article!]!
  }

  type Article {
    id: ID!
    title: String
    author: String
    url: String
    time: String
    source: String
  }
`;
