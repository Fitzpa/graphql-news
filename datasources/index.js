import { HackerNewsAPI } from './hackernews';
import { NewYorkTimesAPI } from './newyorktimes';

export default {
  hackernews: new HackerNewsAPI(),
  newyorktimes: new NewYorkTimesAPI(),
};
