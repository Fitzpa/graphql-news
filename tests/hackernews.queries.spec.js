import { createTestClient } from 'apollo-server-testing';
import { ApolloServer, gql } from 'apollo-server';
import { HackerNewsAPI } from '../datasources/hackernews';
import { resolvers } from '../resolvers';
import { typeDefs } from '../typedefs';
import { getArticlePreReducerStub } from '../fixtures/hackernews';

const GET_ARTICLE_BY_ID_AND_SOURCE = gql`
  query getArticleByIdAndSource($id: ID!, $source: String!) {
    articleBySource(id: $id, source: $source) {
      id
      title
      author
      url
      time
      source
    }
  }
`;

const GET_ARTICLES_BY_SOURCE = gql`
  query getArticlesBySource($ids: [Int!]!, $source: String!) {
    articlesBySource(ids: $ids, source: $source) {
      id
      title
      author
      url
      time
      source
    }
  }
`;

const GET_ALL_ARTICLES = gql`
  query getAllArticles {
    allArticles {
      id
      title
      author
      url
      time
      source
    }
  }
`;

const constructTestServer = () => {
  const hackernewsAPI = new HackerNewsAPI();

  const server = new ApolloServer({
    typeDefs,
    resolvers,
    dataSources: () => ({
      hackernews: hackernewsAPI,
    }),
  });

  return { server, hackernewsAPI };
};

describe('[Queries.HackerNewsAPI]', () => {
  it('fetches an article from the HackerNews API', async () => {
    const { server, hackernewsAPI } = constructTestServer();

    hackernewsAPI.get = jest.fn(() => getArticlePreReducerStub);

    const { query } = createTestClient(server);

    const res = await query({
      query: GET_ARTICLE_BY_ID_AND_SOURCE,
      variables: { id: 21168364, source: 'hackernews' },
    });

    expect(res).toMatchSnapshot();
  });

  it('fetches an array of articles from the HackerNews API', async () => {
    const { server, hackernewsAPI } = constructTestServer();

    hackernewsAPI.get = jest.fn(() => getArticlePreReducerStub);

    const { query } = createTestClient(server);

    const res = await query({
      query: GET_ARTICLES_BY_SOURCE,
      variables: { ids: [21168364], source: 'hackernews' },
    });

    expect(res).toMatchSnapshot();
  });

  it('fetches an array of all the articles from the HackerNews API', async () => {
    const { server, hackernewsAPI } = constructTestServer();

    hackernewsAPI.get = jest.fn(() => [getArticlePreReducerStub]);

    const { query } = createTestClient(server);

    const res = await query({
      query: GET_ALL_ARTICLES,
    });

    expect(res).toMatchSnapshot();
  });
});
