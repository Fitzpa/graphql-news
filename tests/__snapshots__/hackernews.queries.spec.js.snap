// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Queries.HackerNewsAPI] fetches an array of all the articles from the HackerNews API 1`] = `
Object {
  "data": Object {
    "allArticles": Array [
      Object {
        "author": null,
        "id": "hn-undefined",
        "source": "HackerNews",
        "time": null,
        "title": null,
        "url": null,
      },
    ],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`[Queries.HackerNewsAPI] fetches an array of articles from the HackerNews API 1`] = `
Object {
  "data": Object {
    "articlesBySource": Array [
      Object {
        "author": "kristianp",
        "id": "hn-21168364",
        "source": "HackerNews",
        "time": "1570314101",
        "title": "How to Stress the C# Compiler",
        "url": "https://blog.hediet.de/post/how-to-stress-the-csharp-compiler",
      },
    ],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`[Queries.HackerNewsAPI] fetches an article from the HackerNews API 1`] = `
Object {
  "data": Object {
    "articleBySource": Object {
      "author": "kristianp",
      "id": "hn-21168364",
      "source": "HackerNews",
      "time": "1570314101",
      "title": "How to Stress the C# Compiler",
      "url": "https://blog.hediet.de/post/how-to-stress-the-csharp-compiler",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
